Middleware: Don't require HTTPS for HSTS headers to be emitted

alexmv opened the pull request:
Grafana itself may not be serving content over HTTPS, but it may be
behind a transparent proxy which does.

Fixes #26770.  Based on #26868.

alexmv committed:
Middleware: Don't require HTTPS for HSTS headers to be emitted

Grafana itself may not be serving content over HTTPS, but it may be
behind a transparent proxy which does.

Fixes #26770.  Based on #26868.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=35147) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=35147) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=35147) it.</sub>

alexmv commented:
The second paragraph made me real leery as I realized and wrote it.  Fixing this without opening that can of worms likely requires some heuristic on incoming HTTP requests that leads us to believe that they're coming in over HTTPS in some layer -- and I don't think there's any standard way of doing that.

oddlittlebird commented:
Hi Alex,

People are heads-down working on Grafana 8.0 atm, but hopefully a dev can look at this after that!

alexmv commented:
No worries -- thanks for the note.  👍

alexmv commented:
Ping on this?

alexmv commented:
This is a pretty simple PR -- any chance someone can take a look at it?

Thanks!

achatterjee-grafana reviewed (commented):
Looks good to me.

achatterjee-grafana commented:
@wbrowne and @ying-jeanne, can you please review. Thanks!

alexmv commented:
Gentle poke on this, @wbrowne / @ying-jeanne?

alexmv commented:
Actually, I think checking X-Forwarded-Proto is the safe way to actually implement this, since proxies should set that.  I'll reword with that change.

achatterjee-grafana commented:
Hello @wbrowne and @ying-jeanne, can one of you look through the PR? Thanks!

alexmv commented:
Rebased, and I removed the caveat about being problematic because it could make the site inaccessible if it _wasn't_ behind HTTPS.

Because, as [mentioned on developer.mozilla.com](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security#description):
> **Note**: The `Strict-Transport-Security` header is **ignored** by the browser when your site is accessed using HTTP; this is because an attacker may intercept HTTP connections and inject the header or remove it. When your site is accessed over HTTPS with no certificate errors, the browser knows your site is HTTPS capable and will honor the `Strict-Transport-Security` header.

sakjur commented:
@alexmv Thank you for the PR and patience 🙂💪
